File: app/src/main/java/com/example/myapplication/Main_Screen.java
package com.example.myapplication;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.util.TypedValue;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.LinearLayout.LayoutParams;
import android.widget.TextView;
import android.widget.Toast;

import org.w3c.dom.Text;

import java.util.ArrayList;

public class Main_Screen extends AppCompatActivity {
    public static final String BADGE_NUMBER = "com.example.MyApplication.MESSAGE";
    @Override
    public void onUserInteraction() {
        super.onUserInteraction();
        MainActivity.setInactivityFromAnotherActivity(false);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_screen);
        Intent intent = getIntent();
        String message = intent.getStringExtra(MainActivity.EXTRA_MESSAGE);
        TextView title = (TextView)findViewById(R.id.textView2);
        AlertDialog.Builder alertDialog = null;
        title.setText("Main menu for : " + message);
        ArrayList<ArrayList<String>> tcodes = (ArrayList<ArrayList<String>>) getIntent().getSerializableExtra("key");
        LinearLayout layout = (LinearLayout)findViewById(R.id.linearlayout);
        String res = null;
        int btn_id = 0;
        LinearLayout row = new LinearLayout(this);

        for(ArrayList<String> i : tcodes){
            btn_id++;
            String single_function = "";
            String description = null;
            for (int j=0; j < i.size(); j++){
                single_function = i.get(0);
                description = i.get(1);
            }

            if(single_function.equals("P08")){
                String test = "";
            }


            if(btn_id % 3 == 1) {

                row = new LinearLayout(this);
                row.setLayoutParams(new LinearLayout.LayoutParams
                        (LinearLayout.LayoutParams.MATCH_PARENT,
                                LinearLayout.LayoutParams.MATCH_PARENT));
                row.setOrientation(LinearLayout.HORIZONTAL);
                Button button = new Button(this);
                button.setText(description);
                button.setTextSize(TypedValue.COMPLEX_UNIT_PX, 18);
                //Drawable top = getResources().getDrawable(R.drawable.aikitting);
                //button.setCompoundDrawablesWithIntrinsicBounds(null,top,null,null);
                button.setId(btn_id);
                LinearLayout.LayoutParams param = new LinearLayout.LayoutParams(
                        LayoutParams.MATCH_PARENT,
                        LayoutParams.MATCH_PARENT,
                        1.0f
                );
                String classtext = "com.example.myapplication." + single_function;
                button.setLayoutParams(param);
                button.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Button b = (Button)view;
                        String buttonText = b.getText().toString();
                        Toast.makeText(Main_Screen.this, "Showing " + buttonText , Toast.LENGTH_LONG).show();

                        try {
                            Class<?> c = Class.forName(classtext);
                            Intent intent = new Intent(Main_Screen.this, c);
                            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            intent.putExtra(BADGE_NUMBER,message);
                            startActivity(intent);

                        } catch (ClassNotFoundException ignored) {
                            String eror = ignored.getMessage().toString();
                            String error = "";

                        }

                    }
                });
                row.addView(button);
            }
            else if (btn_id % 3 == 2){
                Button button = new Button(this);
                button.setText(description);
                button.setTextSize(TypedValue.COMPLEX_UNIT_PX, 18);
                //Drawable top = getResources().getDrawable(R.drawable.aikitting);
                //button.setCompoundDrawablesWithIntrinsicBounds(null,top,null,null);
                button.setId(btn_id);
                LinearLayout.LayoutParams param = new LinearLayout.LayoutParams(
                        LayoutParams.MATCH_PARENT,
                        LayoutParams.MATCH_PARENT,
                        1.0f
                );

                String classtext = "com.example.myapplication." + single_function;
                button.setLayoutParams(param);
                button.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Button b = (Button)view;
                        String buttonText = b.getText().toString();
                        Toast.makeText(Main_Screen.this, "Showing " + buttonText , Toast.LENGTH_LONG).show();

                        try {
                            Class<?> c = Class.forName(classtext);
                            Intent intent = new Intent(Main_Screen.this, c);
                            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            intent.putExtra(BADGE_NUMBER,message);
                            startActivity(intent);
                        } catch (ClassNotFoundException ignored) {
                            String eror = ignored.getMessage().toString();
                            String error = "";

                        }


                    }
                });
                row.addView(button);
                if(tcodes.size()==2){
                    layout.addView(row);
                }
            }
            else{
                Button button = new Button(this);
                button.setText(description);
                button.setTextSize(TypedValue.COMPLEX_UNIT_PX, 18);
                //Drawable top = getResources().getDrawable(R.drawable.aikitting);
                //button.setCompoundDrawablesWithIntrinsicBounds(null,top,null,null);
                button.setId(btn_id);
                LinearLayout.LayoutParams param = new LinearLayout.LayoutParams(
                        LayoutParams.MATCH_PARENT,
                        LayoutParams.MATCH_PARENT,
                        1.0f
                );

                String classtext = "com.example.myapplication." + single_function;
                button.setLayoutParams(param);
                button.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        Button b = (Button)view;
                        String buttonText = b.getText().toString();
                        Toast.makeText(Main_Screen.this, "Showing " + buttonText , Toast.LENGTH_LONG).show();

                        try {
                            Class<?> c = Class.forName(classtext);
                            Intent intent = new Intent(Main_Screen.this, c);
                            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            intent.putExtra(BADGE_NUMBER,message);
                            startActivity(intent);

                        } catch (ClassNotFoundException ignored) {
                            String eror = ignored.getMessage().toString();
                            String error = "";

                        }

                    }
                });
                row.addView(button);
                layout.addView(row);
            }

            if(tcodes.size()==1){
                layout.addView(row);
            }



        }

    }
}

File: app/src/main/java/com/example/myapplication/Main_Screen.kt
package com.example.myapplication

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TableRow
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import org.json.JSONArray
import org.json.JSONObject

class Main_Screen : AppCompatActivity() {
    private lateinit var mainLayout:LinearLayout
    private lateinit var c:Context
    private lateinit var message: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main_screen)
        c = this
        //val intent = intent
        message = intent.getStringExtra("Badge").toString()

        val jsonString = intent.getStringExtra("Functions")
        val title = findViewById<View>(R.id.textView2) as TextView
        title.text = "Main menu for : $message"
        mainLayout = findViewById(R.id.linearlayout)

        var functionArray = JSONArray(jsonString)


        if(functionArray.length() > 0 ){
            var row = LinearLayout(c)
            setRowProperty(row)

            // Calculate the number of rows needed
            val numRows = (functionArray.length() + 2) / 3

            // Create a 2D list to represent the grid
            val grid = List(numRows) { row ->
                List(3) { col ->
                    val index = row * 3 + col
                    if (index < functionArray.length()) functionArray[index] else ""
                }
            }

            for(rowGrid in grid){
                for (item in rowGrid){
                    if (item != ""){
                        var function = JSONObject(item.toString()).getString("Function")
                        var description = JSONObject(item.toString()).getString("Description")
                        row.addView(addbtn(description,function,message,description))
                    }
                }
                mainLayout.addView(row)
                row = LinearLayout(c)
                setRowProperty(row)
            }
        }


    }


    private fun setRowProperty(v:LinearLayout){
        var rowParams: TableRow.LayoutParams = TableRow.LayoutParams(
            TableRow.LayoutParams.MATCH_PARENT,
            TableRow.LayoutParams.MATCH_PARENT)
        with(rowParams){
            weight = 1F
            setMargins(10,10,10,10)
        }
        with(v){
            gravity = android.view.Gravity.CENTER
            layoutParams = rowParams
            orientation = LinearLayout.HORIZONTAL
            //setBackgroundResource(R.drawable.border)
        }
    }

    private fun addbtn(desc:String,tcd:String,badgeNum:String,des:String):View{
        val funcBtn = Button(c)
        with(funcBtn){
            text = desc
            textSize = 12F
            gravity = Gravity.CENTER
            layoutParams = LinearLayout.LayoutParams(
                200,
                200,
                1.0f
            )
        }

        funcBtn.setOnClickListener{

            try {
                val classText = "com.example.myapplication.$tcd"
                val className = Class.forName(classText)
                var newActivityIntent = Intent(c,className)
                newActivityIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
                newActivityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                newActivityIntent.putExtra("Badge", badgeNum)
                newActivityIntent.putExtra("Desc",des)
                startActivity(newActivityIntent)

            }
            catch (ex:Exception){

            }

        }
        return funcBtn
    }


}